### PR TITLE
Allow users to access stage breakdowns for help

### PR DESCRIPTION
**Pull request by codyduval**
__Description of Change(s) Introduced:__ 

- Adds a question mark icon to the header which, when clicked,
invokes the stage breakdown modal for the current retro stage.

- Adds a new redux action for alerts to invoke help.

- Adds a new 'help' key to the stageConfig.js to show abbreviated
header for the help modal (does not show "Stage Change! when it
is help).

- Since the header exists outside of the react dom, uses a React
portal to render the question icon.

- Quesion mark icon renders conditionally based on stage, only
showing when there is help to be shown.

__Description of Testing Applied:__

Enzyme tests for new `StageHelp` component.

__Relevant github Issue:__ (if applicable)

- Issue #489

### CODE REVIEW

File: web/static/js/configs/stage_configs.jsx [hunk 1]
```diff
@@ -21,6 +21,10 @@ const baseIdeaGenerationConfig = {
     headerText: "Stage Change: Idea Generation!",
     BodyComponent: StageChangeInfoIdeaGeneration,
   },
+  help: {
+    headerText: "Idea Generation",
+    BodyComponent: StageChangeInfoIdeaGeneration,
+  },
```
Comment (codyduval) [inline web/static/js/configs/stage_configs.jsx]: @vanderhoop There were a few ways to solve the "Don't show the Stage Change!" text requirement, I chose to introduce some duplication in the config object instead of some sort of conditional in the component.  Pros and cons with each approach, but I felt was the cleanest.

File: test/redux/alert_test.js [hunk 2]
```diff
@@ -59,6 +59,39 @@ describe("alert", () => {
...
+            bodyText: "Tell me what to do",
+          },
+        },
+        stageLackingHelpConfig: {},
+      }
+
+      deepFreeze(initialState)
+      deepFreeze(stageConfigs)
+
+      describe("when the given stage has a help in the given configuration map", () => {
```
Comment (vanderhoop) [inline test/redux/alert_test.js]: a thought: "do we need a test to handle the other case?"

File: web/static/js/components/stage_help.jsx [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        />
+      </div>,
+      document.getElementById("stage-help-icon")
+    )
+  }
+
+  return null
+}
+
+const mapStateToProps = state => {
```
Comment (vanderhoop) [inline web/static/js/components/stage_help.jsx]: I don't think this component will need to be `connect`ed, as it's parent, the `<RemoteRetro />` component, already has the `retro` state and it can pass it down, right?

File: web/static/js/redux/alert.js [hunk 4]
```diff
@@ -1,5 +1,7 @@
 const types = {
   CLEAR_ALERT: "CLEAR_ALERT",
+  RETRO_UPDATE_COMMITTED: "RETRO_UPDATE_COMMITTED",
```
Comment (vanderhoop) [inline web/static/js/redux/alert.js]: I realize this is a refactor, but if we're going to clean this up, would it be better to import the `types` from the "retro" redux file?

File: web/static/js/prop_types.js [hunk 5]
```diff
@@ -58,3 +58,11 @@ export const votes = PropTypes.arrayOf(PropTypes.object)
 export const ideas = PropTypes.arrayOf(idea)
 
 export const actions = PropTypes.object
+
```
Comment (vanderhoop) [inline web/static/js/prop_types.js]: ❤️ thank you for doing this

File: web/static/js/configs/stage_configs.jsx (hunk 1 above)
Comment (vanderhoop) [reply web/static/js/configs/stage_configs.jsx]: yeah, i hear that. think this is a sensible choice.

File: web/static/js/components/stage_help.jsx [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+export const StageHelp = props => {
+  const handleClick = () => {
+    const { retro, actions: { showStageHelp } } = props
+    showStageHelp(retro)
+  }
+
+  const showIcon = () => {
+    const { retro: { stage } } = props
+    const validHelpStages = [
+      STAGES.PRIME_DIRECTIVE,
```
Comment (vanderhoop) [inline web/static/js/components/stage_help.jsx]: hmmm... this datastructure is implicitly coupled to the stage config object, as the only stage omitted here is the one state where `help` is null. so maybe we could just key into that object and `showIcon` if and only if `!help`?

File: test/components/stage_help_test.js [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+
+describe("<StageHelp />", () => {
+  let wrapper
+  let retro
+
+  // Setup JSDom so that react can inject a portal
+  const iconRoot = global.document.createElement("div")
+  iconRoot.setAttribute("id", "stage-help-icon")
+  const body = global.document.querySelector("body")
+  body.appendChild(iconRoot)
```
Comment (vanderhoop) [inline test/components/stage_help_test.js]: dude, nice work on these tests. have to imagine it took some digging to get the setup right.

File: web/static/js/redux/alert.js (hunk 4 above)
Comment (vanderhoop) [reply web/static/js/redux/alert.js]: and in fact, `SHOW_STAGE_HELP` is also in the `retros` redux file's `types`, so if we *can* import them, I think we should house both there. any reason we wouldn't? not sure if there's a redux convention for this

File: web/static/js/components/stage_help.jsx (hunk 3 above)
Comment (codyduval) [reply web/static/js/components/stage_help.jsx]: Yes, it looks like the `retro` state is in the parent - good catch.  But I still need to `dispatch` from the component so I'll need to `connect`.

File: test/redux/alert_test.js (hunk 2 above)
Comment (codyduval) [reply test/redux/alert_test.js]: Well, the other case will return an `undefined`. I can make the reducer return `null` instead and then the `Alert` component should just not render if it hits that condition.

File: web/static/js/redux/alert.js (hunk 4 above)
Comment (codyduval) [reply web/static/js/redux/alert.js]: Hmm - convention is all over the place here.  Since we're housing reducers and actionCreators together, we should probably split the types.  So SHOW_STAGE_HELP and CLEAR_ALERT stay in the alert redux file, and the others stay in the retros file.  We then import in both files...

File: web/static/js/components/stage_help.jsx (hunk 6 above)
Comment (codyduval) [reply web/static/js/components/stage_help.jsx]: Cool - good idea.

Comment (codyduval) [reply web/static/js/components/stage_help.jsx]: The internet knows all.

File: test/components/stage_help_test.js (hunk 7 above)
Comment (codyduval) [reply test/components/stage_help_test.js]: The internet knows all.

File: test/redux/alert_test.js (hunk 2 above)
Comment (vanderhoop) [reply test/redux/alert_test.js]: i believe the `Alert` component already does a null render when there's no `alert` object passed as a prop, so we're good there

File: web/static/js/components/stage_help.jsx [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+        />
+      </div>,
+      document.getElementById("stage-help-icon")
+    )
+  }
+
+  return null
+}
+
+const mapDispatchToProps = dispatch => ({
```
Comment (vanderhoop) [inline web/static/js/components/stage_help.jsx]: @codyduval the `<RemoteRetro />` component also has bound `actions`, so i think you can kill the `connect` call outright and pass the `actions` object down

Comment (vanderhoop) [reply web/static/js/components/stage_help.jsx]: peep https://github.com/stride-nyc/remote_retro/blob/master/web/static/js/components/remote_retro.jsx#L100-L102

File: web/static/js/components/stage_help.jsx [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+
+export const StageHelp = props => {
+  const handleClick = () => {
+    const { retro, actions: { showStageHelp } } = props
+    showStageHelp(retro)
+  }
+
+  const showIcon = () => {
+    const { retro: { stage } } = props
+    return (stageConfig[stage] && stageConfig[stage].help)
```
Comment (vanderhoop) [inline web/static/js/components/stage_help.jsx]: we should be able to axe the short-circuiting here, as each `stage` is a constant, and are all accounted for in the `stageConfig` object. While in theory they _can_ change, any unaccounted for changes would result in a number of end-to-end tests failing, so we shouldn't need to be defensive here. unless i'm missing something?

File: web/static/js/redux/alert.js [hunk 10]
```diff
@@ -1,5 +1,8 @@
-const types = {
+import { types as retroTypes } from "./retro"
+
+export const types = {
   CLEAR_ALERT: "CLEAR_ALERT",
+  SHOW_STAGE_HELP: "SHOW_STAGE_HELP",
```
Comment (vanderhoop) [inline web/static/js/redux/alert.js]: ❤️

File: web/static/js/components/stage_help.jsx (hunk 8 above)
Comment (codyduval) [reply web/static/js/components/stage_help.jsx]: Hmm - I've never passed an action creator already bound to dispatch.  Lemme see if that works.

File: web/static/js/components/stage_help.jsx (hunk 9 above)
Comment (codyduval) [reply web/static/js/components/stage_help.jsx]: Sure - it's just such a common error I just do it reflexively.

File: test/redux/alert_test.js [hunk 11]
```diff
@@ -59,6 +59,56 @@ describe("alert", () => {
...
+      describe("when the given stage does NOT have a help in the given configuration map", () => {
+        const action = {
+          type: "SHOW_STAGE_HELP",
+          retro: {
+            stage: "nightfall",
+          },
+          stageConfigs,
+        }
+
+        it("returns the help for the given stage", () => {
```
Comment (vanderhoop) [inline test/redux/alert_test.js]: we'll need to update the test language here to reflect that we expect null

File: web/static/js/components/stage_help.jsx (hunk 3 above)
Comment (codyduval) [reply web/static/js/components/stage_help.jsx]: You're right, the actions can get passed down.

File: web/static/js/components/stage_help.jsx (hunk 8 above)
Comment (codyduval) [reply web/static/js/components/stage_help.jsx]: Yup works.

File: test/redux/alert_test.js (hunk 11 above)
Comment (codyduval) [reply test/redux/alert_test.js]: 👍

**Comment by vanderhoop** [comment]
dude, really nice work on this. mergin!
